cpp/include/raft/core/device_resources_manager.hpp

```diff
@@ -0,0 +1,580 @@
+  device_resources_manager(device_resources_manager const&) = delete;
+  void operator=(device_resources_manager const&)           = delete;
+
+ private:
+  device_resources_manager() {}
+  ~device_resources_manager()
+  {
+    // Ensure that we destroy any pool memory resources before CUDA context is
+    // lost
+    per_device_components_.clear();
+  }
+
+  // Get an id used to identify this thread for the purposes of assigning
+  // (in round-robin fashion) the same resources to the thread on subsequent calls to
+  // `get_device_resources`
+  static auto get_thread_id()
+  {
+    static std::atomic<std::size_t> thread_counter{};
+    thread_local std::size_t id = ++thread_counter;
+    return id;
+  }
+
+  // This struct holds the various parameters used to control
+  // construction of the underlying resources shared by all
+  // `device_resources` objects returned by `get_device_resources`
+  struct resource_params {
+    // The total number of primary streams to be used by the
+    // application. If no value is provided, the default stream per thread
+    // is used.
+    std::optional<std::size_t> stream_count{std::nullopt};
+    // The total number of stream pools to be used by the application
+    std::size_t pool_count{};
+    // How many streams to assign to each pool
+    std::size_t pool_size{rmm::cuda_stream_pool::default_size};
+    // If a memory pool is requested (max_mem_pool_size is non-zero), use
+    // this initial size for the pool in bytes. Must be a multiple of 256.
+    // If nullopt, use half of the available memory on the current
+    // device.
+    thrust::optional<std::size_t> init_mem_pool_size{thrust::nullopt};
+    // If set to any non-zero value, create a memory pool with this
+    // maximum size. If nullopt, use up to the entire available memory of the
+    // device
+    thrust::optional<std::size_t> max_mem_pool_size{std::size_t{}};
+    // Limit on workspace memory for the returned device_resources object
+    std::optional<std::size_t> workspace_allocation_limit{std::nullopt};
+  } params_;
+
+  // This struct stores the underlying resources to be shared among
+  // `device_resources` objects returned by this manager.
+  struct resource_components {
+    // Construct all underlying resources indicated by `params` for the
+    // indicated device. This includes primary streams, stream pools, and
+    // a memory pool if requested.
+    resource_components(int device_id, resource_params const& params)
+      : device_id_{device_id},
+        streams_{[&params, this]() {
+          auto scoped_device = device_setter{device_id_};
+          auto result        = std::unique_ptr<rmm::cuda_stream_pool>{nullptr};
+          if (params.stream_count) {
+            result = std::make_unique<rmm::cuda_stream_pool>(*params.stream_count);
+          }
+          return result;
+        }()},
+        pools_{[&params, this]() {
+          auto scoped_device = device_setter{device_id_};
```

Comment:
I think the device id can and should be an instance variable and state on each individual `device_resources` instances and once "set", the corresponding device_resources should be the "source of truth" for the device id it's mapped to. Again- all of the state for each individual device_resources instances should be managed down at that layer- the device_resources_manager should just be managing its own device_resources instances.

Reply:
We still need to generate the streams to be shared by the `device_resources` objects, and those need to be associated with a device id. If we're caching the `device_resources` objects, then yes, we can query those objects in the `find` call to retrieve the `device_resources` associated with the requested device.